subsys/net/lib/sockets/sockets.c

```diff
@@ -1572,21 +1580,28 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 	if (msg != NULL) {
 		if (msg->msg_control != NULL) {
 			if (msg->msg_controllen > 0) {
-				bool clear_controllen = true;
+				bool clear_controllen = false;
 
 				if (IS_ENABLED(CONFIG_NET_CONTEXT_TIMESTAMPING)) {
-					clear_controllen = false;
 					if (add_timestamping(ctx, pkt, msg) < 0) {
 						msg->msg_flags |= ZSOCK_MSG_CTRUNC;
+					} else {
+						clear_controllen = false;
 					}
+				} else {
+					clear_controllen = true;
 				}
 
 				if (IS_ENABLED(CONFIG_NET_CONTEXT_RECV_PKTINFO) &&
 				    net_context_is_recv_pktinfo_set(ctx)) {
 					clear_controllen = false;
 					if (add_pktinfo(ctx, pkt, msg) < 0) {
 						msg->msg_flags |= ZSOCK_MSG_CTRUNC;
+					} else {
+						clear_controllen = false;
 					}
+				} else {
+					clear_controllen = true;
```

Comment:
as an example for my previous comment, if we have both timestamping and pkt_info enabled, we can have a case where we add timestamping info, then there are no pkt_info to add so we go to this branch and we would clear the control data, including the timestamping.